src/passes/Unsubtyping.cpp

```diff
@@ -0,0 +1,580 @@
+  void visitAtomicNotify(AtomicNotify* curr) {}
+  void visitAtomicFence(AtomicFence* curr) {}
+  void visitSIMDExtract(SIMDExtract* curr) {}
+  void visitSIMDReplace(SIMDReplace* curr) {}
+  void visitSIMDShuffle(SIMDShuffle* curr) {}
+  void visitSIMDTernary(SIMDTernary* curr) {}
+  void visitSIMDShift(SIMDShift* curr) {}
+  void visitSIMDLoad(SIMDLoad* curr) {}
+  void visitSIMDLoadStoreLane(SIMDLoadStoreLane* curr) {}
+  void visitMemoryInit(MemoryInit* curr) {}
+  void visitDataDrop(DataDrop* curr) {}
+  void visitMemoryCopy(MemoryCopy* curr) {}
+  void visitMemoryFill(MemoryFill* curr) {}
+  void visitConst(Const* curr) {}
+  void visitUnary(Unary* curr) {}
+  void visitBinary(Binary* curr) {}
+  void visitSelect(Select* curr) {
+    noteSubtype(curr->ifTrue->type, curr->type);
+    noteSubtype(curr->ifFalse->type, curr->type);
+  }
+  void visitDrop(Drop* curr) {}
+  void visitReturn(Return* curr) {
+    if (curr->value) {
+      noteSubtype(curr->value->type, getFunction()->getResults());
+    }
+  }
+  void visitMemorySize(MemorySize* curr) {}
+  void visitMemoryGrow(MemoryGrow* curr) {}
+  void visitUnreachable(Unreachable* curr) {}
+  void visitPop(Pop* curr) {}
+  void visitRefNull(RefNull* curr) {}
+  void visitRefIsNull(RefIsNull* curr) {}
+  void visitRefFunc(RefFunc* curr) {}
+  void visitRefEq(RefEq* curr) {}
+  void visitTableGet(TableGet* curr) {}
+  void visitTableSet(TableSet* curr) {
+    noteSubtype(curr->value->type, getModule()->getTable(curr->table)->type);
+  }
+  void visitTableSize(TableSize* curr) {}
+  void visitTableGrow(TableGrow* curr) {}
+  void visitTableFill(TableFill* curr) {
+    noteSubtype(curr->value->type, getModule()->getTable(curr->table)->type);
+  }
+  void visitTry(Try* curr) {
+    noteSubtype(curr->body->type, curr->type);
+    for (auto* body : curr->catchBodies) {
+      noteSubtype(body->type, curr->type);
+    }
+  }
+  void visitThrow(Throw* curr) {
+    Type params = getModule()->getTag(curr->tag)->sig.params;
+    assert(params.size() == curr->operands.size());
+    for (size_t i = 0, size = curr->operands.size(); i < size; ++i) {
+      noteSubtype(curr->operands[i]->type, params[i]);
+    }
+  }
+  void visitRethrow(Rethrow* curr) {}
+  void visitTupleMake(TupleMake* curr) {}
+  void visitTupleExtract(TupleExtract* curr) {}
+  void visitRefI31(RefI31* curr) {}
+  void visitI31Get(I31Get* curr) {}
+  void visitCallRef(CallRef* curr) {
+    if (!curr->target->type.isSignature()) {
+      return;
+    }
+    handleCall(curr, curr->target->type.getHeapType().getSignature());
+  }
+  void visitRefTest(RefTest* curr) {
+    noteCast(curr->ref->type, curr->castType);
+  }
+  void visitRefCast(RefCast* curr) { noteCast(curr->ref->type, curr->type); }
+  void visitBrOn(BrOn* curr) {
+    if (curr->op == BrOnCast || curr->op == BrOnCastFail) {
+      noteCast(curr->ref->type, curr->castType);
+      noteSubtype(curr->getSentType(), findBreakTarget(curr->name)->type);
```

Comment:
`BrOnNonNull` needs this line as well, I think?
```wat
(block (result $super)
  (br_on_non_null (local.get $sub)) ;; writes a sub to a super slot
)
```
(it's true the br by itself just affects nullability, but the heap type can still be a subtype of the target block)